algorithms/maths/num_perfect_squares.py

```diff
@@ -0,0 +1,41 @@
+"""
+Given an integer num_perfect_squares will return the minimum amount of perfect squares are required
+to sum to the specified number. Lagrange's four-square theorem gives us that the answer will always
+be between 1 and 4 (https://en.wikipedia.org/wiki/Lagrange%27s_four-square_theorem).
+
+Some examples:
+Number | Perfect Squares representation | Answer
+-------|--------------------------------|--------
+9      | 3^2                            | 1
+10     | 3^2 + 1^2                      | 2
+12     | 2^2 + 2^2 + 2^2                | 3
+31     | 5^2 + 2^2 + 1^2 + 1^2          | 4
+"""
+
+import math
+
+def num_perfect_squares(number):
+    """
+    Returns the smallest number of perfect squares that sum to the specified number.
+    :return: int between 1 - 4
+    """
+    # If the number is a perfect square then we only need 1 number.
+    if int(math.sqrt(number))**2 == number:
+        return 1
+
+    # We check if https://en.wikipedia.org/wiki/Legendre%27s_three-square_theorem holds.
+    while number > 0 and number % 4 == 0:
+        number /= 4
+
+    # If it cannot be expressed as three squares it has to be made of four squares
+    # in accordance with https://en.wikipedia.org/wiki/Lagrange%27s_four-square_theorem
+    if number % 8 == 7:
+        return 4
+
+    # Check if there is any number `i` from 1 to sqrt(number) that will result in two 
+    # perfect squares.
+    for i in range(1, int(math.sqrt(number)) + 1):
+        if int(math.sqrt(number - i**2))**2 == number - i**2:
+            return 2
```

Comment:
Should the test for `2` perfect squares be placed before the check for `4`?

Reply:
It should not. Purely functionally it can, however this would lower the performance of the function. I made another commit (c75594e1b499130845550ad04447229fc3003107) which adds more documentation to better explain what is going on and clarify why the code pieces have the order that they have.